sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/ShardListingCheckpointGenerator.java

```diff
@@ -0,0 +1,135 @@
+
+  public ShardListingCheckpointGenerator(KinesisIO.Read spec) {
+    this.spec = spec;
+  }
+
+  @Override
+  public <ClientT> KinesisReaderCheckpoint generate(ClientT client)
+      throws TransientKinesisException {
+    StartingPoint startingPoint = spec.getInitialPosition();
+
+    if (client instanceof SimplifiedKinesisClient) {
+      SimplifiedKinesisClient kinesis = (SimplifiedKinesisClient) client;
+      List<Shard> streamShards =
+          kinesis.listShardsAtPoint(
+              Preconditions.checkArgumentNotNull(spec.getStreamName()),
+              Preconditions.checkArgumentNotNull(startingPoint));
+
+      LOG.info("Creating a checkpoint with following shards {} at {}", streamShards, startingPoint);
+      return new KinesisReaderCheckpoint(
+          streamShards.stream()
+              .map(
+                  shard ->
+                      new ShardCheckpoint(
+                          Preconditions.checkArgumentNotNull(spec.getStreamName()),
+                          shard.shardId(),
+                          Preconditions.checkArgumentNotNull(startingPoint)))
+              .collect(Collectors.toList()));
+    } else if (client instanceof KinesisAsyncClient) {
+      KinesisAsyncClient kinesis = (KinesisAsyncClient) client;
+      List<ShardCheckpoint> streamShards = generateShardsCheckpoints(spec, kinesis);
+      LOG.info("Creating a checkpoint with following shards {} at {}", streamShards, startingPoint);
+      return new KinesisReaderCheckpoint(streamShards);
+    } else {
+      throw new IllegalStateException(String.format("Unknown type of client %s", client));
+    }
+  }
+
+  private static List<ShardCheckpoint> generateShardsCheckpoints(
+      KinesisIO.Read readSpec, KinesisAsyncClient kinesis) {
+    ListShardsRequest listShardsRequest =
+        ListShardsRequest.builder()
+            .streamName(checkArgumentNotNull(readSpec.getStreamName()))
+            .shardFilter(buildFilter(readSpec))
+            .build();
+
+    ListShardsResponse response = tryListingShards(listShardsRequest, kinesis);
+    return response.shards().stream()
+        .map(
+            s ->
+                new ShardCheckpoint(
+                    checkArgumentNotNull(readSpec.getStreamName()),
+                    s.shardId(),
+                    checkArgumentNotNull(readSpec.getInitialPosition())))
+        .collect(Collectors.toList());
+  }
+
+  private static ListShardsResponse tryListingShards(
+      ListShardsRequest listShardsRequest, KinesisAsyncClient kinesis) {
+    try {
+      LOG.info("Starting ListShardsRequest {}", listShardsRequest);
+      ListShardsResponse response =
+          kinesis.listShards(listShardsRequest).get(shardListingTimeoutMs, TimeUnit.MILLISECONDS);
+      LOG.info("Shards found = {}", response.shards());
+      return response;
+    } catch (ExecutionException | InterruptedException | TimeoutException e) {
+      LOG.error("Error listing shards {}", e.getMessage());
```

Comment:
Please don't swallow exceptions, either rethrow, wrap (as cause) or log the exception (and not just the message).
When  testing I misspelled my stream name. This made it unnecessary hard to find the issue.